flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java

```diff
@@ -0,0 +1,562 @@
+        final OneInputStreamOperatorTestHarness<String, Object> testHarness =
+                new OneInputStreamOperatorTestHarness<>(new StreamSink<>(sink));
+
+        testHarness.open();
+
+        // Let the whole bulk request fail
+        responses.add(response -> response.setStatusCode(500));
+
+        testHarness.processElement(new StreamRecord<>("msg"));
+
+        assertThatThrownBy(() -> testHarness.snapshot(1L, 1000L))
+                .getCause()
+                .getCause()
+                .hasMessageContaining("Unable to parse response body");
+    }
+
+    /**
+     * Tests that any bulk failure in the listener callbacks due to flushing on an immediately
+     * following checkpoint is rethrown; we set a timeout because the test will not finish if the
+     * logic is broken.
+     */
+    @Test
+    @Timeout(5)
+    public void testBulkFailureRethrownOnOnCheckpointAfterFlush() throws Throwable {
+        final OpensearchSink.Builder<String> builder =
+                new OpensearchSink.Builder<>(
+                        Arrays.asList(new HttpHost("localhost", server.getLocalPort())),
+                        new SimpleSinkFunction<String>());
+        builder.setBulkFlushInterval(500);
+        builder.setFailureHandler(new NoOpFailureHandler());
+
+        final OpensearchSink<String> sink = builder.build();
+        final OneInputStreamOperatorTestHarness<String, Object> testHarness =
+                new OneInputStreamOperatorTestHarness<>(new StreamSink<>(sink));
+
+        testHarness.open();
+
+        responses.add(
+                createResponse(
+                        new BulkItemResponse(
+                                1,
+                                OpType.INDEX,
+                                new IndexResponse(
+                                        new ShardId("test", "-", 0), "_doc", "1", 0, 0, 1, true))));
+
+        // Let the whole bulk request fail
+        responses.add(response -> response.setStatusCode(500));
+
+        // setup the next bulk request, and let bulk request succeed
+        testHarness.processElement(new StreamRecord<>("msg-1"));
+
+        // Await for flush to be complete
+        awaitForFlushToFinish(1);
+
+        // setup the requests to be flushed in the snapshot
+        testHarness.processElement(new StreamRecord<>("msg-2"));
+        testHarness.processElement(new StreamRecord<>("msg-3"));
+
+        CheckedThread snapshotThread =
+                new CheckedThread() {
+                    @Override
+                    public void go() throws Exception {
+                        testHarness.snapshot(1L, 1000L);
+                    }
+                };
+        snapshotThread.start();
+
+        // Await for flush to be complete
+        awaitForFlushToFinish(0);
+
+        assertThatThrownBy(snapshotThread::sync)
+                .getCause()
+                .getCause()
+                .hasMessageContaining("Unable to parse response body");
```

Comment:
This sounds like an error in the test setup.

Reply:
Not really, so the original test case is designed to verify the condition "complete bulk response fails". The Bulk API (in Elasticsearch or Opensearch) has does not support that, the failure are only part of individual bulk item [1]. The way I was able to replicate this kind of scenario is by sending invalid payload as part of bulk response but I could not control what kind of exception should be raised (hence we have `Unable to parse response body`), I could add comment about it to clarify it is by design.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#docs-bulk-api-example

Reply:
> The Bulk API (in Elasticsearch or Opensearch) does not support that

Couldn't it fail as a whole because of connection issues etc.?

In any case this should be fine then.